src/store_client.cc

```diff
@@ -527,12 +561,19 @@ store_client::readBody(const char *, ssize_t len)
              * copyInto.offset includes headers, which is what mem cache needs
              */
             int64_t mem_offset = entry->mem_obj->endOffset();
-            if ((copyInto.offset == mem_offset) || (parsed_header && mem_offset == rep->hdr_sz)) {
```

Comment:
I speculate that the second condition was incorrectly added by beae59b to compensate for another bug that was later fixed in 619da1e. I suspect the second condition is currently never true in the official and PR code. When you get a chance, please check the offsets and sizes involved in all three cases to confirm or reject my suspicion:

1. Successfully just parsed an HTTP header (parsed_header is true).
2. Successfully parsed an HTTP header in the past. Now adding HTTP body.
3. Need more data to parse an HTTP header.

Reply:
If my suspicion is correct (i.e. the right part of the condition is always false), then I would keep it as is unless we have to modify the condition for some other reason. Armed with the proof requested on this thread, we can remove the bogus condition in a dedicated followup PR.

Reply:
I ran tests against pre-PR Squid confirming that the `(parsed_header && mem_offset == rep->hdr_sz)` condition is never true. Here are my results (with parameter values), for each store_client::readBody() run:

1. **Small headers (<4K) reply**

|readBody() iteration| copyInto.offset | mem_offset | parsed_header | rep->hdr_sz |
| --- | --- | --- | --- | --- |
| 1 | 0 | 0  | 1  | 250 |
| 2 | 4001 | 4001  | 0  | 250 |
| 3 | 12193 | 12193  | 0  | 250 |

2. **Large headers (>4K) reply**

|readBody() iteration| copyInto.offset | mem_offset | parsed_header | rep->hdr_sz |
| --- | --- | --- | --- | --- |
| 1 | 0 | 0  | 0  | 0 |
| 1 | 3967 | 3967  | 0  | 0 |
| 1 | 8063 | 9063  | 0  | 0 |

Though the 'large headers' test resulted in an invalid reply (as already discussed in this issue), but nevertheless, we see that the second condition is false in this case also.

Reply:
> unless we have to modify the condition 

I had to modify this condition. First I thought that it is possible to unconditionally copy the read bytes (copyInto buffer) into the memory (even if the headers are not parsed yet) - this would happen when preserving the `copyInto.offset == mem_offset` condition. However, in this case, when another client starts reading, scheduleRead() will call scheduleMemRead() - and the unparsed header bytes will be transferred back to the client (since scheduleMemRead() just reads copies the from the memory and invokes the callback). So now this method starts copying back to memory after the successful header parsing.

Reply:
> when another client starts reading ... the unparsed header bytes will be transferred back to the client

And that is bad because the client cannot handle partial headers, right? In other words, some or all Store clients expect a complete HTTP header at the beginning of the buffer upon receiving the first "data available" notification from the Store, correct?

Reply:
> upon receiving the first "data available" notification

I don't think that clients wait for 'data available' notification on the initial read operation - the entry is already on disk (i.e., available for hits) and clients get it immediately. The point is that the first client has to load it from the disk while the second one may obtain it from memory (since the first client copies read bytes to the memory). Ather that, a callback is called (e.g., clientReplyContext::cacheHit()) and these bytes are eventually passed on the wire. There are no checks in cacheHit() (or below) that the headers are parsed. The (2) [table](https://github.com/measurement-factory/squid/pull/73#discussion_r434019003) confirms that: unparsed headers (due to the parsing error) and the body are treated as a 'message body' and passed to the client (via an invalid `HTTP/1.1   0 Init` reply).

Reply:
I am answering my own question by combining pieces of Eduard's (correct) answer and additional clarifications:

If we simply write unparsed reply header (prefix) into mem_obj, then the second client that discovers the same StoreEntry, may be immediately notified about hit availability (clientReplyContext::cacheHit() will be called) and:

1. The second client will treat mem_obj::reply_ as parsed/ready-to-use while that reply object (always allocated by MemObject constructor) will be in the middle of its incremental parsing process -- it should not be touched or used except by the parsing code in store_client. This is why Eduard mentioned `HTTP/1.1 0 Init` symptom in his answer -- cloneReply() clones a half-baked reply object and, hence, getting half-baked reply headers to send to the HTTP client.

2. The second client will skip zero bytes (while thinking that it is skipping parsed reply header bytes) and send unparsed reply data after sending half-baked reply headers (from item 1). This is why Eduard says "and these bytes are eventually passed on the wire":
```C++
    flags.headersSent = true;
    ...
    char *body_buf = buf + reply->hdr_sz; // XXX: reply->hdr_sz is still zero so we skip nothing
```

A natural followup question is why put (parsed) reply headers into the buffer if the client is going to skip them anyway (while using/cloning a parsed reply object instead)? I bet that the correct answer is that we should not put them there, but it is a very difficult improvement because many offsets/sizes assume (often implicitly) that the headers are there. Removing them would require identifying all that difficult-to-find code and adjusting it. Such an improvement would deserve a dedicated PR (or ten).

Another followup question is whether some other code might start using half-parsed mem_obj::reply_ for something (while our store_client code is still parsing it). Prior to this PR, that reply object was either completely unparsed or (nearly always successfully) parsed. We are adding a third state. I do not know how Squid code decides that mem_obj::reply_ is usable, but I doubt there is one safe API for checking that. I plan to avoid storing a half-parsed reply in mem_obj to remove this danger.

I am not requesting any code changes. Just documenting my current understanding of this tricky issue.

Reply:
N.B. MemStore::copyFromShmSlice() stores half-parsed reply header bytes into mem_obj, probably breaking things as detailed in my comment above. It also bypasses some checks in Http::Message::parse() by calling httpMsgParseStep() instead; I have not checked why or what problems that might cause.

Reply:
I posted an official PR 1122 to remove this always-false condition. I am still not requesting any code changes or even merge updates for this PR.